File: MyLeetCodeAnswers/C++/ValidParentheses.cpp
class Solution {
public:
    char getRight(char left) {
        switch (left) {
            case '(': return ')';
            case '{': return '}';
            case '[': return ']';
            default: return '\0';
        }
    }
    
    inline bool isLeft(char left) {
        return (left == '(' || left == '{' || left == '[');
    }
    
    bool isValid(string s) {
        stack<char> sta;
        for (int i = 0; i < s.size(); ++i) {
            if (isLeft(s[i])) {
                sta.push(s[i]);
            } else {
                if (sta.empty()) return false;
                else if (getRight(sta.top()) != s[i]) {
                    return false;
                } else {
                    sta.pop();
                }
            }
        }
        return sta.empty();
    }
};

/*Improved Version*/
class Solution {
public:
    char getRight(char left) {
        switch (left) {
            case '(': return ')';
            case '{': return '}';
            case '[': return ']';
            default: return '\0';
        }
    }
    
    inline bool isLeft(char left) {
        return (left == '(' || left == '{' || left == '[');
    }
    
    bool isValid(string s) {
        stack<char> sta;
        for (int i = 0; i < s.size(); ++i) {
            if (isLeft(s[i])) {
                sta.push(s[i]);
            } else {
                if (sta.empty() || getRight(sta.top()) != s[i]) return false;
                else sta.pop();
            }
        }
        return sta.empty();
    }
};

File: MyLeetCodeAnswers/java/ValidParentheses.java
import java.util.Stack;

/**
 * Source   : https://leetcode.com/problems/valid-parentheses/
 * Author   : Wang Meng
 * Date     : 2016-03-18
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 * <p>
 * The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
 */
public class ValidParentheses {
    public boolean isValid(String s) {
        if (s == null || s.isEmpty()) return true;
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); ++i) {
            if (stack.empty() || !isPair(stack.peek(), s.charAt(i))) {
                stack.push(s.charAt(i));
            } else {
                stack.pop();
            }
        }
        return stack.empty();
    }

    private boolean isPair(char a, char b) {
        return (a == '(' && b == ')') || (a == '[' && b == ']') || (a == '{' && b == '}');
    }
}
